// @midscene/shared/dist/es/env/types.mjs
const MIDSCENE_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_MODEL_NAME = 'MIDSCENE_MODEL_NAME';
const MIDSCENE_LANGSMITH_DEBUG = 'MIDSCENE_LANGSMITH_DEBUG';
const MIDSCENE_DEBUG_AI_PROFILE = 'MIDSCENE_DEBUG_AI_PROFILE';
const MIDSCENE_DEBUG_AI_RESPONSE = 'MIDSCENE_DEBUG_AI_RESPONSE';
const MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG = 'MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG';
const MIDSCENE_DEBUG_MODE = 'MIDSCENE_DEBUG_MODE';
const MIDSCENE_MCP_USE_PUPPETEER_MODE = 'MIDSCENE_MCP_USE_PUPPETEER_MODE';
const MIDSCENE_MCP_CHROME_PATH = 'MIDSCENE_MCP_CHROME_PATH';
const MIDSCENE_MCP_ANDROID_MODE = 'MIDSCENE_MCP_ANDROID_MODE';
const DOCKER_CONTAINER = 'DOCKER_CONTAINER';
const MIDSCENE_FORCE_DEEP_THINK = 'MIDSCENE_FORCE_DEEP_THINK';
const MIDSCENE_OPENAI_SOCKS_PROXY = 'MIDSCENE_OPENAI_SOCKS_PROXY';
const MIDSCENE_OPENAI_HTTP_PROXY = 'MIDSCENE_OPENAI_HTTP_PROXY';
const OPENAI_API_KEY = 'OPENAI_API_KEY';
const OPENAI_BASE_URL = 'OPENAI_BASE_URL';
const OPENAI_MAX_TOKENS = 'OPENAI_MAX_TOKENS';
const MIDSCENE_ADB_PATH = 'MIDSCENE_ADB_PATH';
const MIDSCENE_ADB_REMOTE_HOST = 'MIDSCENE_ADB_REMOTE_HOST';
const MIDSCENE_ADB_REMOTE_PORT = 'MIDSCENE_ADB_REMOTE_PORT';
const MIDSCENE_ANDROID_IME_STRATEGY = 'MIDSCENE_ANDROID_IME_STRATEGY';
const MIDSCENE_CACHE = 'MIDSCENE_CACHE';
const MIDSCENE_USE_VLM_UI_TARS = 'MIDSCENE_USE_VLM_UI_TARS';
const MIDSCENE_USE_QWEN_VL = 'MIDSCENE_USE_QWEN_VL';
const MIDSCENE_USE_DOUBAO_VISION = 'MIDSCENE_USE_DOUBAO_VISION';
const MIDSCENE_USE_GEMINI = 'MIDSCENE_USE_GEMINI';
const MIDSCENE_USE_VL_MODEL = 'MIDSCENE_USE_VL_MODEL';
const MATCH_BY_POSITION = 'MATCH_BY_POSITION';
const MIDSCENE_API_TYPE = 'MIDSCENE-API-TYPE';
const MIDSCENE_REPORT_TAG_NAME = 'MIDSCENE_REPORT_TAG_NAME';
const MIDSCENE_REPLANNING_CYCLE_LIMIT = 'MIDSCENE_REPLANNING_CYCLE_LIMIT';
const MIDSCENE_PREFERRED_LANGUAGE = 'MIDSCENE_PREFERRED_LANGUAGE';
const MIDSCENE_USE_AZURE_OPENAI = 'MIDSCENE_USE_AZURE_OPENAI';
const MIDSCENE_AZURE_OPENAI_SCOPE = 'MIDSCENE_AZURE_OPENAI_SCOPE';
const MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_CACHE_MAX_FILENAME_LENGTH = 'MIDSCENE_CACHE_MAX_FILENAME_LENGTH';
const AZURE_OPENAI_ENDPOINT = 'AZURE_OPENAI_ENDPOINT';
const AZURE_OPENAI_KEY = 'AZURE_OPENAI_KEY';
const AZURE_OPENAI_API_VERSION = 'AZURE_OPENAI_API_VERSION';
const AZURE_OPENAI_DEPLOYMENT = 'AZURE_OPENAI_DEPLOYMENT';
const MIDSCENE_USE_ANTHROPIC_SDK = 'MIDSCENE_USE_ANTHROPIC_SDK';
const ANTHROPIC_API_KEY = 'ANTHROPIC_API_KEY';
const MIDSCENE_RUN_DIR = 'MIDSCENE_RUN_DIR';
const MIDSCENE_OPENAI_BASE_URL = 'MIDSCENE_OPENAI_BASE_URL';
const MIDSCENE_OPENAI_API_KEY = 'MIDSCENE_OPENAI_API_KEY';
const MIDSCENE_OPENAI_USE_AZURE = 'MIDSCENE_OPENAI_USE_AZURE';
const MIDSCENE_AZURE_OPENAI_KEY = 'MIDSCENE_AZURE_OPENAI_KEY';
const MIDSCENE_AZURE_OPENAI_ENDPOINT = 'MIDSCENE_AZURE_OPENAI_ENDPOINT';
const MIDSCENE_AZURE_OPENAI_API_VERSION = 'MIDSCENE_AZURE_OPENAI_API_VERSION';
const MIDSCENE_AZURE_OPENAI_DEPLOYMENT = 'MIDSCENE_AZURE_OPENAI_DEPLOYMENT';
const MIDSCENE_ANTHROPIC_API_KEY = 'MIDSCENE_ANTHROPIC_API_KEY';
const MIDSCENE_VL_MODE = 'MIDSCENE_VL_MODE';
const MIDSCENE_VQA_MODEL_NAME = 'MIDSCENE_VQA_MODEL_NAME';
const MIDSCENE_VQA_OPENAI_SOCKS_PROXY = 'MIDSCENE_VQA_OPENAI_SOCKS_PROXY';
const MIDSCENE_VQA_OPENAI_HTTP_PROXY = 'MIDSCENE_VQA_OPENAI_HTTP_PROXY';
const MIDSCENE_VQA_OPENAI_BASE_URL = 'MIDSCENE_VQA_OPENAI_BASE_URL';
const MIDSCENE_VQA_OPENAI_API_KEY = 'MIDSCENE_VQA_OPENAI_API_KEY';
const MIDSCENE_VQA_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_VQA_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_VQA_OPENAI_USE_AZURE = 'MIDSCENE_VQA_OPENAI_USE_AZURE';
const MIDSCENE_VQA_USE_AZURE_OPENAI = 'MIDSCENE_VQA_USE_AZURE_OPENAI';
const MIDSCENE_VQA_AZURE_OPENAI_SCOPE = 'MIDSCENE_VQA_AZURE_OPENAI_SCOPE';
const MIDSCENE_VQA_AZURE_OPENAI_KEY = 'MIDSCENE_VQA_AZURE_OPENAI_KEY';
const MIDSCENE_VQA_AZURE_OPENAI_ENDPOINT = 'MIDSCENE_VQA_AZURE_OPENAI_ENDPOINT';
const MIDSCENE_VQA_AZURE_OPENAI_API_VERSION = 'MIDSCENE_VQA_AZURE_OPENAI_API_VERSION';
const MIDSCENE_VQA_AZURE_OPENAI_DEPLOYMENT = 'MIDSCENE_VQA_AZURE_OPENAI_DEPLOYMENT';
const MIDSCENE_VQA_AZURE_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_VQA_AZURE_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_VQA_USE_ANTHROPIC_SDK = 'MIDSCENE_VQA_USE_ANTHROPIC_SDK';
const MIDSCENE_VQA_ANTHROPIC_API_KEY = 'MIDSCENE_VQA_ANTHROPIC_API_KEY';
const MIDSCENE_VQA_VL_MODE = 'MIDSCENE_VQA_VL_MODE';
const MIDSCENE_PLANNING_MODEL_NAME = 'MIDSCENE_PLANNING_MODEL_NAME';
const MIDSCENE_PLANNING_OPENAI_SOCKS_PROXY = 'MIDSCENE_PLANNING_OPENAI_SOCKS_PROXY';
const MIDSCENE_PLANNING_OPENAI_HTTP_PROXY = 'MIDSCENE_PLANNING_OPENAI_HTTP_PROXY';
const MIDSCENE_PLANNING_OPENAI_BASE_URL = 'MIDSCENE_PLANNING_OPENAI_BASE_URL';
const MIDSCENE_PLANNING_OPENAI_API_KEY = 'MIDSCENE_PLANNING_OPENAI_API_KEY';
const MIDSCENE_PLANNING_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_PLANNING_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_PLANNING_OPENAI_USE_AZURE = 'MIDSCENE_PLANNING_OPENAI_USE_AZURE';
const MIDSCENE_PLANNING_USE_AZURE_OPENAI = 'MIDSCENE_PLANNING_USE_AZURE_OPENAI';
const MIDSCENE_PLANNING_AZURE_OPENAI_SCOPE = 'MIDSCENE_PLANNING_AZURE_OPENAI_SCOPE';
const MIDSCENE_PLANNING_AZURE_OPENAI_KEY = 'MIDSCENE_PLANNING_AZURE_OPENAI_KEY';
const MIDSCENE_PLANNING_AZURE_OPENAI_ENDPOINT = 'MIDSCENE_PLANNING_AZURE_OPENAI_ENDPOINT';
const MIDSCENE_PLANNING_AZURE_OPENAI_API_VERSION = 'MIDSCENE_PLANNING_AZURE_OPENAI_API_VERSION';
const MIDSCENE_PLANNING_AZURE_OPENAI_DEPLOYMENT = 'MIDSCENE_PLANNING_AZURE_OPENAI_DEPLOYMENT';
const MIDSCENE_PLANNING_AZURE_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_PLANNING_AZURE_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_PLANNING_USE_ANTHROPIC_SDK = 'MIDSCENE_PLANNING_USE_ANTHROPIC_SDK';
const MIDSCENE_PLANNING_ANTHROPIC_API_KEY = 'MIDSCENE_PLANNING_ANTHROPIC_API_KEY';
const MIDSCENE_PLANNING_VL_MODE = 'MIDSCENE_PLANNING_VL_MODE';
const MIDSCENE_GROUNDING_MODEL_NAME = 'MIDSCENE_GROUNDING_MODEL_NAME';
const MIDSCENE_GROUNDING_OPENAI_SOCKS_PROXY = 'MIDSCENE_GROUNDING_OPENAI_SOCKS_PROXY';
const MIDSCENE_GROUNDING_OPENAI_HTTP_PROXY = 'MIDSCENE_GROUNDING_OPENAI_HTTP_PROXY';
const MIDSCENE_GROUNDING_OPENAI_BASE_URL = 'MIDSCENE_GROUNDING_OPENAI_BASE_URL';
const MIDSCENE_GROUNDING_OPENAI_API_KEY = 'MIDSCENE_GROUNDING_OPENAI_API_KEY';
const MIDSCENE_GROUNDING_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_GROUNDING_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_GROUNDING_OPENAI_USE_AZURE = 'MIDSCENE_GROUNDING_OPENAI_USE_AZURE';
const MIDSCENE_GROUNDING_USE_AZURE_OPENAI = 'MIDSCENE_GROUNDING_USE_AZURE_OPENAI';
const MIDSCENE_GROUNDING_AZURE_OPENAI_SCOPE = 'MIDSCENE_GROUNDING_AZURE_OPENAI_SCOPE';
const MIDSCENE_GROUNDING_AZURE_OPENAI_KEY = 'MIDSCENE_GROUNDING_AZURE_OPENAI_KEY';
const MIDSCENE_GROUNDING_AZURE_OPENAI_ENDPOINT = 'MIDSCENE_GROUNDING_AZURE_OPENAI_ENDPOINT';
const MIDSCENE_GROUNDING_AZURE_OPENAI_API_VERSION = 'MIDSCENE_GROUNDING_AZURE_OPENAI_API_VERSION';
const MIDSCENE_GROUNDING_AZURE_OPENAI_DEPLOYMENT = 'MIDSCENE_GROUNDING_AZURE_OPENAI_DEPLOYMENT';
const MIDSCENE_GROUNDING_AZURE_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_GROUNDING_AZURE_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_GROUNDING_USE_ANTHROPIC_SDK = 'MIDSCENE_GROUNDING_USE_ANTHROPIC_SDK';
const MIDSCENE_GROUNDING_ANTHROPIC_API_KEY = 'MIDSCENE_GROUNDING_ANTHROPIC_API_KEY';
const MIDSCENE_GROUNDING_VL_MODE = 'MIDSCENE_GROUNDING_VL_MODE';
const OPENAI_USE_AZURE = 'OPENAI_USE_AZURE';
const UNUSED_ENV_KEYS = [
    MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG
];
const BASIC_ENV_KEYS = [
    MIDSCENE_DEBUG_MODE,
    MIDSCENE_DEBUG_AI_PROFILE,
    MIDSCENE_DEBUG_AI_RESPONSE,
    MIDSCENE_RUN_DIR
];
const BOOLEAN_ENV_KEYS = [
    MIDSCENE_CACHE,
    MIDSCENE_LANGSMITH_DEBUG,
    MIDSCENE_FORCE_DEEP_THINK,
    MIDSCENE_MCP_USE_PUPPETEER_MODE,
    MIDSCENE_MCP_ANDROID_MODE
];
const NUMBER_ENV_KEYS = [
    MIDSCENE_CACHE_MAX_FILENAME_LENGTH,
    MIDSCENE_REPLANNING_CYCLE_LIMIT
];
const STRING_ENV_KEYS = [
    OPENAI_MAX_TOKENS,
    MIDSCENE_ADB_PATH,
    MIDSCENE_ADB_REMOTE_HOST,
    MIDSCENE_ADB_REMOTE_PORT,
    MIDSCENE_ANDROID_IME_STRATEGY,
    MIDSCENE_REPORT_TAG_NAME,
    MIDSCENE_PREFERRED_LANGUAGE,
    MATCH_BY_POSITION,
    MIDSCENE_MCP_CHROME_PATH,
    DOCKER_CONTAINER
];
const GLOBAL_ENV_KEYS = [
    ...BOOLEAN_ENV_KEYS,
    ...NUMBER_ENV_KEYS,
    ...STRING_ENV_KEYS
];
const MODEL_ENV_KEYS = [
    MIDSCENE_MODEL_NAME,
    MIDSCENE_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_OPENAI_API_KEY,
    MIDSCENE_OPENAI_BASE_URL,
    MIDSCENE_OPENAI_USE_AZURE,
    MIDSCENE_OPENAI_SOCKS_PROXY,
    MIDSCENE_OPENAI_HTTP_PROXY,
    MIDSCENE_USE_AZURE_OPENAI,
    MIDSCENE_AZURE_OPENAI_SCOPE,
    MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_USE_ANTHROPIC_SDK,
    MIDSCENE_USE_VLM_UI_TARS,
    MIDSCENE_USE_QWEN_VL,
    MIDSCENE_USE_DOUBAO_VISION,
    MIDSCENE_USE_GEMINI,
    MIDSCENE_USE_VL_MODEL,
    ANTHROPIC_API_KEY,
    MIDSCENE_AZURE_OPENAI_ENDPOINT,
    MIDSCENE_AZURE_OPENAI_KEY,
    MIDSCENE_AZURE_OPENAI_API_VERSION,
    MIDSCENE_AZURE_OPENAI_DEPLOYMENT,
    MIDSCENE_VL_MODE,
    OPENAI_API_KEY,
    OPENAI_BASE_URL,
    OPENAI_USE_AZURE,
    ANTHROPIC_API_KEY,
    AZURE_OPENAI_ENDPOINT,
    AZURE_OPENAI_KEY,
    AZURE_OPENAI_API_VERSION,
    AZURE_OPENAI_DEPLOYMENT,
    MIDSCENE_VQA_MODEL_NAME,
    MIDSCENE_VQA_OPENAI_SOCKS_PROXY,
    MIDSCENE_VQA_OPENAI_HTTP_PROXY,
    MIDSCENE_VQA_OPENAI_BASE_URL,
    MIDSCENE_VQA_OPENAI_API_KEY,
    MIDSCENE_VQA_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_VQA_OPENAI_USE_AZURE,
    MIDSCENE_VQA_USE_AZURE_OPENAI,
    MIDSCENE_VQA_AZURE_OPENAI_SCOPE,
    MIDSCENE_VQA_AZURE_OPENAI_KEY,
    MIDSCENE_VQA_AZURE_OPENAI_ENDPOINT,
    MIDSCENE_VQA_AZURE_OPENAI_API_VERSION,
    MIDSCENE_VQA_AZURE_OPENAI_DEPLOYMENT,
    MIDSCENE_VQA_AZURE_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_VQA_USE_ANTHROPIC_SDK,
    MIDSCENE_VQA_ANTHROPIC_API_KEY,
    MIDSCENE_VQA_VL_MODE,
    MIDSCENE_PLANNING_MODEL_NAME,
    MIDSCENE_PLANNING_OPENAI_SOCKS_PROXY,
    MIDSCENE_PLANNING_OPENAI_HTTP_PROXY,
    MIDSCENE_PLANNING_OPENAI_BASE_URL,
    MIDSCENE_PLANNING_OPENAI_API_KEY,
    MIDSCENE_PLANNING_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_PLANNING_OPENAI_USE_AZURE,
    MIDSCENE_PLANNING_USE_AZURE_OPENAI,
    MIDSCENE_PLANNING_AZURE_OPENAI_SCOPE,
    MIDSCENE_PLANNING_AZURE_OPENAI_KEY,
    MIDSCENE_PLANNING_AZURE_OPENAI_ENDPOINT,
    MIDSCENE_PLANNING_AZURE_OPENAI_API_VERSION,
    MIDSCENE_PLANNING_AZURE_OPENAI_DEPLOYMENT,
    MIDSCENE_PLANNING_AZURE_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_PLANNING_USE_ANTHROPIC_SDK,
    MIDSCENE_PLANNING_ANTHROPIC_API_KEY,
    MIDSCENE_PLANNING_VL_MODE,
    MIDSCENE_GROUNDING_MODEL_NAME,
    MIDSCENE_GROUNDING_OPENAI_SOCKS_PROXY,
    MIDSCENE_GROUNDING_OPENAI_HTTP_PROXY,
    MIDSCENE_GROUNDING_OPENAI_BASE_URL,
    MIDSCENE_GROUNDING_OPENAI_API_KEY,
    MIDSCENE_GROUNDING_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_GROUNDING_OPENAI_USE_AZURE,
    MIDSCENE_GROUNDING_USE_AZURE_OPENAI,
    MIDSCENE_GROUNDING_AZURE_OPENAI_SCOPE,
    MIDSCENE_GROUNDING_AZURE_OPENAI_KEY,
    MIDSCENE_GROUNDING_AZURE_OPENAI_ENDPOINT,
    MIDSCENE_GROUNDING_AZURE_OPENAI_API_VERSION,
    MIDSCENE_GROUNDING_AZURE_OPENAI_DEPLOYMENT,
    MIDSCENE_GROUNDING_AZURE_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_GROUNDING_USE_ANTHROPIC_SDK,
    MIDSCENE_GROUNDING_ANTHROPIC_API_KEY,
    MIDSCENE_GROUNDING_VL_MODE
];
const ALL_ENV_KEYS = [
    ...UNUSED_ENV_KEYS,
    ...BASIC_ENV_KEYS,
    ...GLOBAL_ENV_KEYS,
    ...MODEL_ENV_KEYS
];
export { ALL_ENV_KEYS, ANTHROPIC_API_KEY, AZURE_OPENAI_API_VERSION, AZURE_OPENAI_DEPLOYMENT, AZURE_OPENAI_ENDPOINT, AZURE_OPENAI_KEY, BASIC_ENV_KEYS, BOOLEAN_ENV_KEYS, DOCKER_CONTAINER, GLOBAL_ENV_KEYS, MATCH_BY_POSITION, MIDSCENE_ADB_PATH, MIDSCENE_ADB_REMOTE_HOST, MIDSCENE_ADB_REMOTE_PORT, MIDSCENE_ANDROID_IME_STRATEGY, MIDSCENE_ANTHROPIC_API_KEY, MIDSCENE_API_TYPE, MIDSCENE_AZURE_OPENAI_API_VERSION, MIDSCENE_AZURE_OPENAI_DEPLOYMENT, MIDSCENE_AZURE_OPENAI_ENDPOINT, MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_AZURE_OPENAI_KEY, MIDSCENE_AZURE_OPENAI_SCOPE, MIDSCENE_CACHE, MIDSCENE_CACHE_MAX_FILENAME_LENGTH, MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG, MIDSCENE_DEBUG_AI_PROFILE, MIDSCENE_DEBUG_AI_RESPONSE, MIDSCENE_DEBUG_MODE, MIDSCENE_FORCE_DEEP_THINK, MIDSCENE_GROUNDING_ANTHROPIC_API_KEY, MIDSCENE_GROUNDING_AZURE_OPENAI_API_VERSION, MIDSCENE_GROUNDING_AZURE_OPENAI_DEPLOYMENT, MIDSCENE_GROUNDING_AZURE_OPENAI_ENDPOINT, MIDSCENE_GROUNDING_AZURE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_GROUNDING_AZURE_OPENAI_KEY, MIDSCENE_GROUNDING_AZURE_OPENAI_SCOPE, MIDSCENE_GROUNDING_MODEL_NAME, MIDSCENE_GROUNDING_OPENAI_API_KEY, MIDSCENE_GROUNDING_OPENAI_BASE_URL, MIDSCENE_GROUNDING_OPENAI_HTTP_PROXY, MIDSCENE_GROUNDING_OPENAI_INIT_CONFIG_JSON, MIDSCENE_GROUNDING_OPENAI_SOCKS_PROXY, MIDSCENE_GROUNDING_OPENAI_USE_AZURE, MIDSCENE_GROUNDING_USE_ANTHROPIC_SDK, MIDSCENE_GROUNDING_USE_AZURE_OPENAI, MIDSCENE_GROUNDING_VL_MODE, MIDSCENE_LANGSMITH_DEBUG, MIDSCENE_MCP_ANDROID_MODE, MIDSCENE_MCP_CHROME_PATH, MIDSCENE_MCP_USE_PUPPETEER_MODE, MIDSCENE_MODEL_NAME, MIDSCENE_OPENAI_API_KEY, MIDSCENE_OPENAI_BASE_URL, MIDSCENE_OPENAI_HTTP_PROXY, MIDSCENE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_OPENAI_SOCKS_PROXY, MIDSCENE_OPENAI_USE_AZURE, MIDSCENE_PLANNING_ANTHROPIC_API_KEY, MIDSCENE_PLANNING_AZURE_OPENAI_API_VERSION, MIDSCENE_PLANNING_AZURE_OPENAI_DEPLOYMENT, MIDSCENE_PLANNING_AZURE_OPENAI_ENDPOINT, MIDSCENE_PLANNING_AZURE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_PLANNING_AZURE_OPENAI_KEY, MIDSCENE_PLANNING_AZURE_OPENAI_SCOPE, MIDSCENE_PLANNING_MODEL_NAME, MIDSCENE_PLANNING_OPENAI_API_KEY, MIDSCENE_PLANNING_OPENAI_BASE_URL, MIDSCENE_PLANNING_OPENAI_HTTP_PROXY, MIDSCENE_PLANNING_OPENAI_INIT_CONFIG_JSON, MIDSCENE_PLANNING_OPENAI_SOCKS_PROXY, MIDSCENE_PLANNING_OPENAI_USE_AZURE, MIDSCENE_PLANNING_USE_ANTHROPIC_SDK, MIDSCENE_PLANNING_USE_AZURE_OPENAI, MIDSCENE_PLANNING_VL_MODE, MIDSCENE_PREFERRED_LANGUAGE, MIDSCENE_REPLANNING_CYCLE_LIMIT, MIDSCENE_REPORT_TAG_NAME, MIDSCENE_RUN_DIR, MIDSCENE_USE_ANTHROPIC_SDK, MIDSCENE_USE_AZURE_OPENAI, MIDSCENE_USE_DOUBAO_VISION, MIDSCENE_USE_GEMINI, MIDSCENE_USE_QWEN_VL, MIDSCENE_USE_VLM_UI_TARS, MIDSCENE_USE_VL_MODEL, MIDSCENE_VL_MODE, MIDSCENE_VQA_ANTHROPIC_API_KEY, MIDSCENE_VQA_AZURE_OPENAI_API_VERSION, MIDSCENE_VQA_AZURE_OPENAI_DEPLOYMENT, MIDSCENE_VQA_AZURE_OPENAI_ENDPOINT, MIDSCENE_VQA_AZURE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_VQA_AZURE_OPENAI_KEY, MIDSCENE_VQA_AZURE_OPENAI_SCOPE, MIDSCENE_VQA_MODEL_NAME, MIDSCENE_VQA_OPENAI_API_KEY, MIDSCENE_VQA_OPENAI_BASE_URL, MIDSCENE_VQA_OPENAI_HTTP_PROXY, MIDSCENE_VQA_OPENAI_INIT_CONFIG_JSON, MIDSCENE_VQA_OPENAI_SOCKS_PROXY, MIDSCENE_VQA_OPENAI_USE_AZURE, MIDSCENE_VQA_USE_ANTHROPIC_SDK, MIDSCENE_VQA_USE_AZURE_OPENAI, MIDSCENE_VQA_VL_MODE, MODEL_ENV_KEYS, NUMBER_ENV_KEYS, OPENAI_API_KEY, OPENAI_BASE_URL, OPENAI_MAX_TOKENS, OPENAI_USE_AZURE, STRING_ENV_KEYS, UNUSED_ENV_KEYS };
