// @midscene/shared/dist/es/constants/index.mjs
import { PLAYWRIGHT_EXAMPLE_CODE, YAML_EXAMPLE_CODE } from "./example-code.mjs";
const TEXT_SIZE_THRESHOLD = 9;
const TEXT_MAX_SIZE = 40;
const CONTAINER_MINI_HEIGHT = 3;
const CONTAINER_MINI_WIDTH = 3;
var constants_NodeType = /*#__PURE__*/ function(NodeType) {
    NodeType["CONTAINER"] = "CONTAINER Node";
    NodeType["FORM_ITEM"] = "FORM_ITEM Node";
    NodeType["BUTTON"] = "BUTTON Node";
    NodeType["A"] = "Anchor Node";
    NodeType["IMG"] = "IMG Node";
    NodeType["TEXT"] = "TEXT Node";
    NodeType["POSITION"] = "POSITION Node";
    return NodeType;
}({});
const PLAYGROUND_SERVER_PORT = 5800;
const SCRCPY_SERVER_PORT = 5700;
const DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT = 5000;
const DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT = 2000;
const DEFAULT_WAIT_FOR_NETWORK_IDLE_TIME = 300;
const DEFAULT_WAIT_FOR_NETWORK_IDLE_CONCURRENCY = 2;
export { CONTAINER_MINI_HEIGHT, CONTAINER_MINI_WIDTH, DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT, DEFAULT_WAIT_FOR_NETWORK_IDLE_CONCURRENCY, DEFAULT_WAIT_FOR_NETWORK_IDLE_TIME, DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT, constants_NodeType as NodeType, PLAYGROUND_SERVER_PORT, PLAYWRIGHT_EXAMPLE_CODE, SCRCPY_SERVER_PORT, TEXT_MAX_SIZE, TEXT_SIZE_THRESHOLD, YAML_EXAMPLE_CODE };
